File: _includes/code/howto/search.similarity.py
# Howto: semantic search - Python examples

# ================================
# ===== INSTANTIATION-COMMON =====
# ================================

import weaviate
import weaviate.classes as wvc
import os

client = weaviate.connect_to_wcs(
    cluster_url=os.getenv("WCS_DEMO_URL"),
    auth_credentials=weaviate.auth.AuthApiKey(os.getenv("WCS_DEMO_RO_KEY")),
    headers={
        "X-OpenAI-Api-Key": os.getenv("OPENAI_APIKEY"),
        "X-Cohere-Api-Key": os.getenv("COHERE_APIKEY"),
    },
)

try:


    # ===============================================
    # ===== QUERY WITH TARGET VECTOR & nearText =====
    # ===============================================

    # NamedVectorNearTextPython
    import weaviate.classes as wvc

    reviews = client.collections.get("WineReviewNV")
    response = reviews.query.near_text(
        query="a sweet German white wine",
        limit=2,
        # highlight-start
        target_vector="title_country",  # Specify the target vector for named vector collections
        # highlight-end
        return_metadata=wvc.query.MetadataQuery(distance=True)
    )

    for o in response.objects:
        print(o.properties)
        print(o.metadata.distance)
    # END NamedVectorNearTextPython

    # Test results
    assert response.objects[0].collection == "WineReviewNV"
    assert len(response.objects) == 2
    assert "title" in response.objects[0].properties.keys()
    assert response.objects[0].metadata.distance is not None
    # End test


    # ===============================
    # ===== QUERY WITH nearText =====
    # ===============================

    # https://weaviate.io/developers/weaviate/api/graphql/search-operators#neartext

    # GetNearTextPython
    import weaviate.classes as wvc

    jeopardy = client.collections.get("JeopardyQuestion")
    # highlight-start
    response = jeopardy.query.near_text(
        query="animals in movies",
    # highlight-end
        limit=2,
        return_metadata=wvc.query.MetadataQuery(distance=True)
    )

    for o in response.objects:
        print(o.properties)
        print(o.metadata.distance)
    # END GetNearTextPython

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    assert len(response.objects) == 2
    assert "question" in response.objects[0].properties.keys()
    assert response.objects[0].metadata.distance is not None
    # End test


    # =================================
    # ===== QUERY WITH nearObject =====
    # =================================

    # https://weaviate.io/developers/weaviate/api/graphql/search-operators#nearobject

    # GetNearObjectPython
    import weaviate.classes as wvc

    jeopardy = client.collections.get("JeopardyQuestion")
    # highlight-start
    response = jeopardy.query.near_object(
        near_object="56b9449e-65db-5df4-887b-0a4773f52aa7",
    # highlight-end
        limit=2,
        return_metadata=wvc.query.MetadataQuery(distance=True)
    )

    for o in response.objects:
        print(o.properties)
        print(o.metadata.distance)
    # END GetNearObjectPython

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    assert len(response.objects) == 2
    assert "question" in response.objects[0].properties.keys()
    assert response.objects[0].metadata.distance is not None
    # End test


    # =================================
    # ===== QUERY WITH nearVector =====
    # =================================

    # https://weaviate.io/developers/weaviate/api/graphql/search-operators#nearvector

    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.fetch_objects(limit=1, include_vector=True)
    query_vector = response.objects[0].vector["default"]

    # GetNearVectorPython
    import weaviate.classes as wvc

    jeopardy = client.collections.get("JeopardyQuestion")
    # highlight-start
    response = jeopardy.query.near_vector(
        near_vector=query_vector, # your query vector goes here
    # highlight-end
        limit=2,
        return_metadata=wvc.query.MetadataQuery(distance=True)
    )

    for o in response.objects:
        print(o.properties)
        print(o.metadata.distance)
    # END GetNearVectorPython

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    assert len(response.objects) == 2
    assert "question" in response.objects[0].properties.keys()
    assert response.objects[0].metadata.distance is not None
    # End test


    # ==========================================
    # ===== QUERY WITH OFFSET AND LIMIT =====
    # ==========================================


    # Query with LimitOffset - https://weaviate.io/developers/weaviate/api/graphql/filters#limit-argument

    # GetLimitOffsetPython
    import weaviate.classes as wvc

    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.near_text(
        query="animals in movies",
        # highlight-start
        limit=2,  # return 2 objects
        offset=1,  # With an offset of 1
        # highlight-end
        return_metadata=wvc.query.MetadataQuery(distance=True)
    )

    for o in response.objects:
        print(o.properties)
        print(o.metadata.distance)
    # END GetLimitOffsetPython

    # Test results
    no_offset_response = jeopardy.query.near_text(
        query="animals in movies",
        limit=3,
        return_metadata=wvc.query.MetadataQuery(distance=True)
    )

    assert response.objects[0].collection == "JeopardyQuestion"
    assert len(response.objects) == 2
    assert "question" in response.objects[0].properties.keys()
    assert response.objects[0].metadata.distance is not None
    assert no_offset_response.objects[1].properties["question"] == response.objects[0].properties["question"]
    # End test


    # ===============================
    # ===== QUERY WITH DISTANCE =====
    # ===============================

    # http://weaviate.io/developers/weaviate/config-refs/distances

    # GetWithDistancePython
    import weaviate.classes as wvc

    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.near_text(
        query="animals in movies",
        # highlight-start
        distance=0.18, # max accepted distance
        # highlight-end
        return_metadata=wvc.query.MetadataQuery(distance=True)
    )

    for o in response.objects:
        print(o.properties)
        print(o.metadata.distance)
    # END GetWithDistancePython

    # Test results
    assert response.objects[0].collection == "JeopardyQuestion"
    assert "question" in response.objects[0].properties.keys()
    assert response.objects[0].metadata.distance is not None
    for o in response.objects:
        assert o.metadata.distance < 0.18
    # End test


    # ===============================
    # ===== Query with autocut =====
    # ===============================

    # http://weaviate.io/developers/weaviate/api/graphql/additional-operators#autocut

    # START Autocut Python
    import weaviate.classes as wvc

    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.near_text(
        query="animals in movies",
        # highlight-start
        auto_limit=1, # number of close groups
        # highlight-end
        return_metadata=wvc.query.MetadataQuery(distance=True)
    )

    for o in response.objects:
        print(o.properties)
        print(o.metadata.distance)
    # END Autocut Python

    # Test results
    assert len(response.objects) > 0
    assert response.objects[0].collection == "JeopardyQuestion"
    assert "question" in response.objects[0].properties.keys()
    assert response.objects[0].metadata.distance is not None
    # End test


    # ==============================
    # ===== QUERY WITH groupBy =====
    # ==============================


    # https://weaviate.io/developers/weaviate/api/graphql/get#get-groupby
    # GetWithGroupbyPython
    import weaviate.classes as wvc

    jeopardy = client.collections.get("JeopardyQuestion")
    # highlight-start

    group_by = wvc.query.GroupBy(
        prop="round",  # group by this property
        objects_per_group=2,  # maximum objects per group
        number_of_groups=2,  # maximum number of groups
    )

    response = jeopardy.query.near_text(
        query="animals in movies", # find object based on this query
        limit=10,  # maximum total objects
        return_metadata=wvc.query.MetadataQuery(distance=True),
        group_by=group_by
    )
    # highlight-end


    for o in response.objects:
        print(o.uuid)
        print(o.belongs_to_group)
        print(o.metadata.distance)

    for grp, grp_items in response.groups.items():
        print("=" * 10 + grp_items.name + "=" * 10)
        print(grp_items.number_of_objects)
        for o in grp_items.objects:
            print(o.properties)
            print(o.metadata)
    # END GetWithGroupbyPython

    # Test results
    assert len(response.objects) > 0
    assert response.objects[0].collection == "JeopardyQuestion"
    assert "question" in response.objects[0].properties.keys()
    assert response.objects[0].metadata.distance is not None

    assert len(response.groups) > 0
    assert len(response.groups) <= 2
    for grp, grp_items in response.groups.items():
        assert grp_items.number_of_objects <= 2
    # End test


    # ============================
    # ===== QUERY WITH WHERE =====
    # ============================

    # https://weaviate.io/developers/weaviate/api/graphql/search-operators#neartext

    # GetWithWherePython
    import weaviate.classes as wvc

    jeopardy = client.collections.get("JeopardyQuestion")
    response = jeopardy.query.near_text(
        query="animals in movies",
        # highlight-start
        filters=wvc.query.Filter.by_property("round").equal("Double Jeopardy!"),
        # highlight-end
        limit=2,
        return_metadata=wvc.query.MetadataQuery(distance=True),
    )

    for o in response.objects:
        print(o.properties)
        print(o.metadata.distance)
    # END GetWithWherePython

    # Test results
    assert len(response.objects) > 0
    assert response.objects[0].collection == "JeopardyQuestion"
    assert response.objects[0].properties["round"] == "Double Jeopardy!"
    assert "question" in response.objects[0].properties.keys()
    assert response.objects[0].metadata.distance is not None
    # End test

finally:
    client.close()

File: _includes/code/howto/search.similarity.ts
// How-to: Search -> Similarity search - TypeScript examples

import assert from 'assert';

// ================================
// ===== INSTANTIATION-COMMON =====
// ================================

import weaviate, { WeaviateClient } from 'weaviate-client';

const client: WeaviateClient = await weaviate.connectToWeaviateCloud(
  process.env.WCD_URL,
 {
   authCredentials: new weaviate.ApiKey(process.env.WCD_API_KEY),
   headers: {
     'X-OpenAI-Api-Key': process.env.OPENAI_APIKEY,  // Replace with your inference API key
   }
 } 
)

// add let result later
// GetNearText  // GetNearVector  // GetNearObject  // GetLimitOffset  // GetWithDistance  // START Autocut  // GetWithGroupBy  // GetWithFilter 
const jeopardy = client.collections.get('JeopardyQuestion');
// END GetNearText  // END GetNearVector  // END GetNearObject  // END GetLimitOffset  // END GetWithDistance  // END Autocut  // END GetWithGroupBy  // END GetWithFilter 


// ===============================================
// ===== QUERY WITH TARGET VECTOR & nearText =====
// ===============================================
{
// NamedVectorNearText
const myNVCollection = client.collections.get('WineReviewNV');

const result = await myNVCollection.query.nearText(['a sweet German white wine'],{
  limit: 2,
  // highlight-start
  targetVector: 'title_country',
  // highlight-end
  returnMetadata: ['distance']
})

for (let object of result.objects) {
  console.log(JSON.stringify(object.properties, null, 2));
  console.log(JSON.stringify(object.metadata?.distance, null, 2));
}
// END NamedVectorNearText

// Tests
// assert.deepEqual(result.objects.length, 2);
}

// =========================
// ===== With NearText =====
// =========================
{
// https://weaviate.io/developers/weaviate/api/graphql/search-operators#neartext
// GetNearText

// highlight-start
const result = await jeopardy.query.nearText('animals in movies', {
// highlight-end
  limit: 2,
  returnMetadata: ['distance']
})

result.objects.forEach(item => {
  console.log(JSON.stringify(item.properties, null, 2))
  console.log(item.metadata?.distance)
})
// END GetNearText

// Tests
// let questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(questionKeys, new Set(['question', 'answer']));
// assert.deepEqual(result.objects.length, 2);
}

// ================================
// ===== With NearVector =====
// ================================
{
const tempResp = await jeopardy.query.fetchObjects({limit:1, includeVector:true})
const queryVector = tempResp.objects[0].vectors.default

// https://weaviate.io/developers/weaviate/api/graphql/search-operators#neartext
// GetNearVector

// highlight-start
const result = await jeopardy.query.nearVector(queryVector, {
// highlight-end
  limit: 2,
  returnMetadata: ['distance']
})

for (let object of result.objects) {
  console.log(JSON.stringify(object.properties, null, 2));
  console.log(JSON.stringify(object.metadata?.distance, null, 2));
}
// END GetNearVector

// Tests
// let questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(questionKeys, new Set(['question', 'answer']));
// assert.deepEqual(result.objects.length, 2);
}

// ================================
// ===== With NearObject =====
// ================================
{
// https://weaviate.io/developers/weaviate/api/graphql/search-operators#nearobject
// GetNearObject

// highlight-start
const result = await jeopardy.query.nearObject('56b9449e-65db-5df4-887b-0a4773f52aa7', {
// highlight-end
  limit: 2,
  returnMetadata: ['distance']
})

for (let object of result.objects) {
  console.log(JSON.stringify(object.properties, null, 2));
  console.log(JSON.stringify(object.metadata?.distance, null, 2));
}
// END GetNearObject

// Tests
// let questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(questionKeys, new Set(['question', 'answer']));
// assert.deepEqual(result.objects.length, 2);
}

{
// Limit - https://weaviate.io/developers/weaviate/api/graphql/filters#limit-argument
// GetLimitOffset

const result = await jeopardy.query.nearText('animals in movies', {
  // highlight-start
  limit: 2,
  offset: 1,
  // highlight-end
  returnMetadata: ['distance']
})

console.log(JSON.stringify(result.objects, null, 2));
// END GetLimitOffset

// Tests
// let questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(questionKeys, new Set(['question', 'answer']));
// assert.deepEqual(result.objects.length, 2);
}

// =========================
// ===== With Distance =====
// =========================
{
// Distance - http://weaviate.io/developers/weaviate/config-refs/distances
// GetWithDistance
// highlight-start
const maxDistance = 0.18;
// highlight-end

const result = await jeopardy.query.nearText('animals in movies', {
  // highlight-start
  distance: maxDistance,
  returnMetadata: ['distance'],
  // highlight-end
})

result.objects.forEach(item => {
  console.log(JSON.stringify(item.properties, null, 2))
  console.log(item.metadata?.distance)
})
// END GetWithDistance
}
// Tests
// questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(questionKeys, new Set(['question', 'answer']));


// ========================
// ===== With Autocut =====
// ========================

// # http://weaviate.io/developers/weaviate/api/graphql/additional-operators#autocut
{
// START Autocut

const result = await jeopardy.query.nearText('animals in movies', {
  // highlight-start
  autoLimit: 1,
  // highlight-end
  returnMetadata: ['distance'],
})

result.objects.forEach(item => {
  console.log(JSON.stringify(item.properties, null, 2))
  console.log(item.metadata?.distance)
})
// END Autocut

// Tests
// questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(questionKeys, new Set(['question', 'answer']));
}

// ========================
// ===== With GroupBy =====
// ========================
{
// groupBy - https://weaviate.io/developers/weaviate/api/graphql/get#get-groupby
// GetWithGroupBy

const result = await jeopardy.query.nearText('animals in movies', {
  limit: 10,
  returnMetadata: ['distance'],
  // highlight-start
  groupBy: {
    property: 'round',
    objectsPerGroup: 2,
    numberOfGroups: 2
  }
  // highlight-end
})

for (let group of result.objects) {
  console.log(group.uuid);
  console.log(JSON.stringify(group.belongsToGroup, null, 2));
  console.log(group.metadata?.distance);
}

// END GetWithGroupBy

// Tests
// questionKeys = new Set(Object.keys(result.objects[0].properties['_additional']));
// assert.deepEqual(questionKeys, new Set(['group']));
// assert.deepEqual(result.objects[0].properties['_additional']['group']['hits'].length, 2);
// assert.deepEqual(result.objects.length, 2);
}

// =======================
// ===== With Filter =====
// =======================

// GetWithFilter

const result = await jeopardy.query.nearText('animals in movies', {
  limit: 2,
  returnMetadata: ['distance'],
  // highlight-start
  filters: jeopardy.filter.byProperty('round').equal('Double Jeopardy!')
  // highlight-end
})

result.objects.forEach(item => {
  console.log(JSON.stringify(item.properties, null, 2))
  console.log(item.metadata?.distance)
})
// END GetWithFilter

// Tests
// questionKeys = new Set(Object.keys(result.objects[0].properties));
// assert.deepEqual(questionKeys, new Set(['question', 'answer', 'round']));
// assert.deepEqual(result.objects[0].properties['round'], 'Double Jeopardy!');
// assert.deepEqual(result.objects.length, 2);
